codespan-reporting/src/term/renderer.rs

```diff
@@ -472,9 +472,34 @@ impl<'writer, 'config> Renderer<'writer, 'config> {
         let space_len = self.config.width(&source[..range.start]);
         write!(self, " {space: >width$}", space = "", width = space_len)?;
         self.set_color(self.styles().label(severity, label_style))?;
+
+        // Pad out the range if the does not end on a unicode char boundary.
+        fn padded_boundary<'a>(range: Range<usize>, source: &'a str) -> Range<usize> {
+            if source.is_char_boundary(range.end) {
+                return range;
+            }
+
+            let total_len = source.len();
+            // because is_char_boundary is true if range.end == source.len()
+            // incrementing this by 1, and doing zero iterations of the following loop
+            // should still result in a string where is_char_boundary returns true.
+            let mut pos = range.end + 1;
+            while pos < total_len {
+                if source.is_char_boundary(pos) {
+                    return range.start..pos;
+                }
+                pos += 1;
+            }
+
+            return range.start..pos;
+        }
```

Comment:
I wonder if we could extract this into a private, local function, and give it some tests? I'm thinking something like:

```rust
fn closest_char_boundary(s: &'a str, byte_index: usize) -> usize {
    ⋮
}

fn slice_at_char_boundaries<'a>(s: &'a str, range: Range<usize>) -> &'a str {
    let start = closest_char_boundary(s, range.start);
    let end = closest_char_boundary(s, range.end);
    &s[start..end]
}
```

I had similar issues with trying to make `files::column_index` look nice, but eventually ended up with something that I thought was pretty good:

https://github.com/brendanzab/codespan/blob/44a88ee938cfbdff0f7ea84b86b256b6aa5c4aab/codespan-reporting/src/files.rs#L126-L156

Perhaps it could be worth looking at it for some ideas on how to clean this up? If not, it's not a big deal, I can always have a look later.

Reply:
I moved them to a file utils.rs, noticing that there are a couple of direct slicings in e.g. view.rs.

This is going to need another round, I came up with a couple cleaner solutions,
one is combinator-esque and the one I felt was a little clearer just uses a for loop.
I left the combinator one commented out in case that is preferred.